File: booking.php
<?php
header("Content-Type: application/json");
header("Access-Control-Allow-Origin: *");
header("Access-Control-Allow-Methods: POST, GET, OPTIONS");
header("Access-Control-Allow-Headers: Content-Type, Authorization");

// Database credentials
$servername = "localhost";
$username = "root";
$password = "";
$dbname = "shri_durga";

// Create connection
$conn = new mysqli($servername, $username, $password, $dbname);

// Check connection
if ($conn->connect_error) {
    die(json_encode(["error" => "Database connection failed: " . $conn->connect_error]));
}

// Get the JSON input
$data = json_decode(file_get_contents("php://input"), true);

// Validate input
if (!$data) {
    echo json_encode(["error" => "Invalid input"]);
    exit;
}

// Assign data to variables with default null check
$dt = $data['dt'] ?? null;
$fname = $data['fname'] ?? null;
$mname = $data['mname'] ?? null;
$lname = $data['lname'] ?? null;
$ph = $data['ph'] ?? null;
$executive_name = $data['executive_name'] ?? null;
$manager_name = $data['manager_name'] ?? null;
$tel_name = $data['tel_name'] ?? null;
$p_mode = $data['p_mode'] ?? null;
$remark = $data['remark'] ?? null;
$nxt_dt = $data['nxt_dt'] ?? null;
$p_name = $data['p_name'] ?? null;
$sqft = $data['sqft'] ?? null;
$rate = $data['rate'] ?? null;
$book_amt = $data['book_amt'] ?? null;
$tamount = $data['tamount'] ?? null;
$balance = $data['balance'] ?? 0; // Default to 0 if not provided
$flt = $data['flt'] ?? 0;         // Default to 0 if not provided
$developer = $data['developer'] ?? null;
$mail = $data['mail'] ?? null;
$dob = $data['dob'] ?? null;
$occupation = $data['occupation'] ?? null;
$adhar = $data['adhar'] ?? null;
$nam = $data['nam'] ?? null;
$relation = $data['relation'] ?? null;
$raddress = $data['raddress'] ?? null;
$paddress = $data['paddress'] ?? null;
$layout = $data['layout'] ?? null;

// Check required fields are not empty
$required_fields = ['dt', 'fname', 'ph', 'executive_name', 'manager_name', 'tel_name', 'rate', 'book_amt', 'tamount', 'nxt_dt', 'sqft', 'p_name'];

foreach ($required_fields as $field) {
    if (empty($data[$field])) {
        echo json_encode(["error" => "Field '$field' is required"]);
        exit;
    }
}

// Prepare statements for each table
$stmt_booking = $conn->prepare("
    INSERT INTO booking_item_master 
    (dt, fname, mname, lname, ph, executive_name, manager_name, tel_name, p_mode, remark, nxt_dt, p_name, sqft, rate, book_amt, tamount) 
    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
");
$stmt_booking->bind_param("ssssssssssssssss", $dt, $fname, $mname, $lname, $ph, $executive_name, $manager_name, $tel_name, $p_mode, $remark, $nxt_dt, $p_name, $sqft, $rate, $book_amt, $tamount);

$stmt_agent = $conn->prepare("
    INSERT INTO agent_master 
    (executive_name, manager_name, tel_name, fname, p_name, rate, book_amt, tamount, flt, dt) 
    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
");
$stmt_agent->bind_param("ssssssssss", $executive_name, $manager_name, $tel_name, $fname, $p_name, $rate, $book_amt, $tamount, $flt, $dt);

$stmt_balance = $conn->prepare("
    INSERT INTO balance_master 
    (fname, p_name, rate, book_amt, tamount, balance, nxt_dt, ph) 
    VALUES (?, ?, ?, ?, ?, ?, ?, ?)
");
$stmt_balance->bind_param("ssssssss", $fname, $p_name, $rate, $book_amt, $tamount, $balance, $nxt_dt, $ph);

$stmt_customer = $conn->prepare("
    INSERT INTO booking_customer 
    (developer, fname, mname, lname, ph, mail, dob, occupation, adhar, nam, relation, raddress, paddress, executive_name, manager_name, tel_name, p_mode, remark, layout, nxt_dt, p_name) 
    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
");
$stmt_customer->bind_param("sssssssssssssssssssss", $developer, $fname, $mname, $lname, $ph, $mail, $dob, $occupation, $adhar, $nam, $relation, $raddress, $paddress, $executive_name, $manager_name, $tel_name, $p_mode, $remark, $layout, $nxt_dt, $p_name);

// Transaction and execution as before
$conn->begin_transaction();
try {
    if (!$stmt_booking->execute()) throw new Exception("Error inserting into booking_item_master: " . $stmt_booking->error);
    if (!$stmt_agent->execute()) throw new Exception("Error inserting into agent_master: " . $stmt_agent->error);
    if (!$stmt_balance->execute()) throw new Exception("Error inserting into balance_master: " . $stmt_balance->error);
    if (!$stmt_customer->execute()) throw new Exception("Error inserting into booking_customer: " . $stmt_customer->error);

    $conn->commit();
    echo json_encode(["message" => "Form submitted successfully"]);
} catch (Exception $e) {
    $conn->rollback();
    echo json_encode(["error" => $e->getMessage()]);
}

// Close statements and connection
$stmt_booking->close();
$stmt_agent->close();
$stmt_balance->close();
$stmt_customer->close();
$conn->close();
?>

File: booking.dart
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';

class SiteBookingPage extends StatefulWidget {
  const SiteBookingPage({super.key});

  @override
  _SiteBookingPageState createState() => _SiteBookingPageState();
}

class _SiteBookingPageState extends State<SiteBookingPage> {
  final Map<String, TextEditingController> controllers = {
    'developer': TextEditingController(),
    'dt': TextEditingController(),
    'fname': TextEditingController(),
    'mname': TextEditingController(),
    'lname': TextEditingController(),
    'ph': TextEditingController(),
    'mail': TextEditingController(),
    'dob': TextEditingController(),
    'occupation': TextEditingController(),
    'adhar': TextEditingController(),
    'nam': TextEditingController(),
    'relation': TextEditingController(),
    'raddress': TextEditingController(),
    'paddress': TextEditingController(),
    'executive_name': TextEditingController(),
    'manager_name': TextEditingController(),
    'tel_name': TextEditingController(),
    'p_mode': TextEditingController(),
    'remark': TextEditingController(),
    'layout': TextEditingController(),
    'nxt_dt': TextEditingController(),
    'p_name': TextEditingController(),
    'sqft': TextEditingController(),
    'rate': TextEditingController(),
    'book_amt': TextEditingController(),
    'tamount': TextEditingController(),
  };

  final requiredFields = [
    'developer',
    'dt',
    'fname',
    'lname',
    'ph',
    'mail',
    'dob',
    'occupation',
    'adhar',
    'raddress',
    'paddress',
    'p_name',
    'sqft',
    'rate',
    'book_amt',
    'tamount',
  ];

  @override
  void dispose() {
    controllers.forEach((_, controller) => controller.dispose());
    super.dispose();
  }

  Future<void> submitForm() async {
    if (!_validateFormWithHighlight()) {
      _showSnackbar(context, 'Error', 'Please fill all required fields');
      return;
    }

    try {
      final formData = controllers.map(
        (key, controller) {
          final value = controller.text.trim();
          return MapEntry(key, value.isNotEmpty ? value : null);
        },
      )..removeWhere((key, value) => value == null);

      final response = await http.post(
        Uri.parse('https://qa.shridurgamanagement.in/sp.php'),
        headers: {'Content-Type': 'application/json'},
        body: json.encode(formData),
      );

      final responseData = json.decode(response.body);

      if (response.statusCode == 200 && responseData['message'] != null) {
        _showSnackbar(context, 'Success', responseData['message']);
      } else {
        _showSnackbar(
            context, 'Error', responseData['error'] ?? 'Submission failed');
      }
    } catch (e) {
      _showSnackbar(context, 'Error', 'Network error: $e');
    }
  }

  bool _validateFormWithHighlight() {
    bool isValid = true;

    for (String key in requiredFields) {
      if (controllers[key]?.text.trim().isEmpty ?? true) {
        setState(() {
          controllers[key]?.text = ''; // Reset the field
        });
        isValid = false;
      }
    }

    return isValid;
  }

  void _showSnackbar(BuildContext context, String title, String message) {
    final snackBar = SnackBar(
      content: Text('$title: $message'),
      backgroundColor: title == 'Error' ? Colors.redAccent : Colors.green,
      behavior: SnackBarBehavior.floating,
      margin: const EdgeInsets.all(16),
    );
    ScaffoldMessenger.of(context).showSnackBar(snackBar);
  }

  String _getLabelText(String fieldName) {
    switch (fieldName) {
      case 'developer':
        return 'Developer Name';
      case 'dt':
        return 'Date';
      case 'fname':
        return 'First Name';
      case 'mname':
        return 'Middle Name';
      case 'lname':
        return 'Last Name';
      case 'ph':
        return 'Phone';
      case 'mail':
        return 'Email';
      case 'dob':
        return 'Date of Birth';
      case 'occupation':
        return 'Occupation';
      case 'adhar':
        return 'Aadhar Number';
      case 'nam':
        return 'Name of Person';
      case 'relation':
        return 'Relation with Developer';
      case 'raddress':
        return 'Residential Address';
      case 'paddress':
        return 'Permanent Address';
      case 'executive_name':
        return 'Executive Name';
      case 'manager_name':
        return 'Manager Name';
      case 'tel_name':
        return 'Telephone Number';
      case 'p_mode':
        return 'Payment Mode';
      case 'remark':
        return 'Remarks';
      case 'layout':
        return 'Layout';
      case 'nxt_dt':
        return 'Next Date';
      case 'p_name':
        return 'Project Name';
      case 'sqft':
        return 'Area in Sqft';
      case 'rate':
        return 'Rate per Sqft';
      case 'book_amt':
        return 'Booking Amount';
      case 'tamount':
        return 'Total Amount';
      default:
        return fieldName;
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text(
          'Booking',
          style: TextStyle(fontWeight: FontWeight.bold, color: Colors.white),
        ),
        backgroundColor: const Color(0xFF800000),
      ),
      body: SingleChildScrollView(
        padding: const EdgeInsets.all(16),
        child: Column(
          children: [
            ...controllers.entries.map(
              (entry) => _buildTextField(entry.key, entry.value),
            ),
            const SizedBox(height: 20),
            ElevatedButton(
              onPressed: submitForm,
              style: ElevatedButton.styleFrom(
                backgroundColor: Colors.blue[800],
                padding:
                    const EdgeInsets.symmetric(vertical: 16, horizontal: 32),
                shape: RoundedRectangleBorder(
                  borderRadius: BorderRadius.circular(30),
                ),
              ),
              child: const Text(
                'Submit',
                style: TextStyle(fontSize: 18),
              ),
            ),
          ],
        ),
      ),
    );
  }

  Widget _buildTextField(String label, TextEditingController controller) {
    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 12.0),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(
            _getLabelText(label),
            style: TextStyle(
              fontWeight: FontWeight.w500,
              fontSize: 16,
              color: Colors.blue[800],
            ),
          ),
          const SizedBox(height: 6),
          TextField(
            controller: controller,
            decoration: InputDecoration(
              border: OutlineInputBorder(
                borderRadius: BorderRadius.circular(12),
                borderSide: BorderSide(color: Colors.grey.shade400),
              ),
              focusedBorder: OutlineInputBorder(
                borderRadius: BorderRadius.circular(12),
                borderSide: BorderSide(color: Colors.blue[800]!),
              ),
              contentPadding:
                  const EdgeInsets.symmetric(vertical: 16, horizontal: 12),
            ),
            style: const TextStyle(fontSize: 16),
          ),
        ],
      ),
    );
  }
}
